File: Semester1/FundProgramming/HomeWorksMenuFP.py
import os 

def thankyou():
    clear_screen()
    red = "\033[1;31m"
    reset = "\033[0m"
    print(red + r"  _______ _                 _                         ")
    print(red + r" |__   __| |               | |                        ")
    print(red + r"    | |  | |__   __ _ _ __ | | __  _   _  ___  _   _  ")
    print(red + r"    | |  |  _ \ / _` | '_ \| |/ / | | | |/ _ \| | | | ")
    print(red + r"    | |  | | | | (_| | | | |   <  | |_| | (_) | |_| | ")
    print(red + r"    |_|  |_| |_|\__,_|_| |_|_|\_\  \__, |\___/ \____| ")
    print(red + r"                                    __/ |             ")
    print(red + r"                                   |___/              " + reset)

def clear_screen():
    os.system('cls' if os.name == 'nt' else 'clear')
    
def GasStation():
    clear_screen()
    DKm = float(input("Enter the distance traveled in kilometers: "))
    PGas = float(input("Enter the price per galon of gasoline: "))
    MonGas = float(input("Enter the total amount of money spent on gasoline: "))
    TimeH, TimeM = map(float, input("Enter the time in hours and minutes (use spaces to separate): ").split())
    TimeT = TimeH + (TimeM/60)
    GalKm = (MonGas/PGas)/DKm
    PromKmH = DKm/TimeT
    PromMS = (DKm*1000)/(TimeT*3600)
    print(f"Spent {GalKm:.2f} galon per kilometer traveled")
    print(f"Its average speed was {PromKmH:.2f}km/h")
    print(f"Its average speed was {PromMS:.2f}m/s")
    input("Press Enter to exit...")

def Triangle():
    clear_screen()
    i = 0
    Sides = []
    for i in range(3):
        side = float(input(f"Enter the {i+1} side of the triangle you wish to create: "))
        Sides.append(side)

    Sides.sort()
    if Sides[0] + Sides[1] > Sides[2]:
        print("The triangle can be formed")
    else:
        print("The triangle can't be formed")
    input("Press Enter to exit...")

def SalaryFoot():
    clear_screen()
    incre = 0
    Fsalary = 0
    salary = float(input("Enter the football player's current salary: "))
    if salary <= 6000:
        incre = 20
        Fsalary = salary + (salary*(incre/100))
    elif salary <=7900:
        incre = 10
        Fsalary = salary + (salary*(incre/100))
    elif salary <= 12000:
        incre = 5
        Fsalary = salary + (salary*(incre/100))
    else:
        incre = 0
        Fsalary = salary
    print(f"Current Salary: {salary}\nIncrease: {incre}%\nFinal Salary: {Fsalary}")
    input("Press Enter to exit...")

def GradeHigh():
    clear_screen()
    j=0
    i=0
    Gstds = []
    stds = int(input("Enter the students they passed: "))
    if stds == 0:
        print("Nobody passed")
    else:
        while i != stds:
            clear_screen()
            grade = float(input("Enter the grades of the students who passed: "))
            if grade > 10 or grade < 0:
                print("Enter a valid grade")
                input("Press Enter to continue...")
            else:
                Gstds.append(grade)
                i += 1
        i = 0
        j = sum(1 for grade in Gstds if grade > 9)
    clear_screen()
    print(f"The {(j*100)/stds}% passed with more than 9")
    input("Press Enter to exit...")

def invernum():
    clear_screen()
    num_str = input("Enter a number: ")
    num = list(num_str)  # Convert string to list to allow changing characters
    i = 0
    j = len(num) - 1

    while i < j:
        # Pythonic swap: We don't need an 'aux' variable!
        num[i], num[j] = num[j], num[i] 
        i += 1
        j -= 1
    inverse_num = "".join(num)
    print(f"The inverse number is: {inverse_num}")
  
    #? Pythonic Method
    # num_str = input("Enter a number: ")
    # [::-1] tells Python to slice the string backwards!
    # inverse_num = num_str[::-1] 
    # print(f"The inverse number is: {inverse_num}")

    input("Press Enter to continue...")

def parking():
    while True:
        clear_screen()
        time = 0
        pay = 0
        print('According to your vehicle, enter:\n"a" to automobile\n"b" for bus\n"t" for truck ')
        opt = input('Enter the vehicle type: ')
        match opt:
            case 'a':
                clear_screen()
                pay = 1.20
                time = float(input('Enter the amount of time in hours that the vehicle spent in the parking lot: '))
                print(f"You must pay: {pay*time}$ ")
                input('Press enter to exit...')
                return
            case 'b':
                clear_screen()
                pay = 2
                time = float(input('Enter the amount of time in hours that the vehicle spent in the parking lot: '))
                print(f"You must pay: {pay*time}$ ")
                input('Press enter to exit...')
                return
            case 't':
                clear_screen()
                pay = 3
                time = float(input('Enter the amount of time in hours that the vehicle spent in the parking lot: '))
                print(f"You must pay: {pay*time}$ ")
                input('Press enter to exit...')
                return
            case _:
                print('Enter a valid option')
                input('Press enter to continue...')

def descount():
    clear_screen()
    tpay = 0
    amou = 0
    price = 0
    while amou == 0:
        clear_screen()
        amou = int(input("Enter the quantity of product you are going to buy: "))
        if amou == 0:
            print("Buy something")
            input("Press enter to exit...")
    while price == 0:
        clear_screen()
        price = float(input("Enter the unit price: "))
        if price == 0:
            print("Enter the correct value")
            input("Press enter to exit...")
    if amou <= 9:
        disc = 0
        tpay = (amou*price)
    elif amou <= 49:
        disc = 3
        tpay =(amou*price) - ((amou*price)*(disc/100))
    elif amou <= 99:
        disc = 5
        tpay =(amou*price) - ((amou*price)*(disc/100))
    elif amou >= 100:
        desc = 10
        tpay =(amou*price) - ((amou*price)*(desc/100))
    print(f"The quantity of product you are going to buy is: {amou} ")
    print(f"The unit price: {price} ")
    print(f"The discount is of the: {disc}%")
    print(f"Your total pay is: {tpay:.2f}")
    input("Press enter to exit...")

def cashier():
    clear_screen()
    money = 0
    denominations = [100, 50, 20, 10, 5, 1]
    while money == 0:
        money = int(input(("Enter the amount of money you are going to withdraw (whole number): ")))
        if money == 0:
            print('Enter a valid number')
            input('Press enter to continue...')
    print("You will receive:")
    for bill in denominations:
        # divmod(money, bill) returns a tuple: (money // bill, money % bill)
        count, money = divmod(money, bill) 
        if count > 0:
            print(f"{count} ${bill} bill(s)")
    input("Press intro to exit...")

def aminum():
    clear_screen()
    num1 = int(input("Enter the first number: "))
    num2 = int(input("Enter the second number: "))
    sum1 = 0
    for i in range(1, num1):
        if num1 % i == 0:
            sum1 += i
    sum2 = 0
    for i in range(1, num2):
        if num2 % i == 0:
            sum2 += i
    if sum1 == num2 and sum2 == num1:
        print("They are amicable numbers")
    else:
        print("They are not amicable numbers")
    input("Press Enter to exit...")

def menu():
    while True:
        clear_screen()
        print("=====PROGRAM MENU=====")
        print("1. Gas Station Program ")
        print("2. Form a triangle")
        print("3. Salary increase in football")
        print("4. Grades higher than 9")
        print("5. Inverse sorter of numbers")
        print("6. Paying for parking")
        print("7. Wholesale discount")
        print("8. Cashier Program")
        print("9. Numbers Friends")
        print("E. Exit")
        option = input("Select an option: ")
        match option:
            case "1":
                GasStation()
            case "2":
                Triangle()
            case "3":
                SalaryFoot()
            case "4":
                GradeHigh()
            case "5":
                invernum()
            case "6":
                parking()
            case "7":
                descount()
            case "8":
                cashier()
            case "9":
                aminum()
            case "E":
                return
            case _:
                clear_screen()
                print("Select a correct option")
                input("Press intro to continue...")

def main():
    clear_screen()
    menu()
    thankyou()
    input("Press intro to exit...")

main()

File: Semester1/FundProgramming/HomeworksMenuFP.java
package AcademicPortfolio.Semester1.FundProgramming;

import java.util.Scanner;
import java.util.Arrays;
import java.util.ArrayList;

public class HomeworksMenuFP {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        menu(input);
        input.close();
    }

    public static void clearScreen() {
        System.out.print("\033[H\033[2J");
        System.out.flush();
    }

    public static void pause(Scanner input) {
        input.nextLine();
        System.out.printf("%nPress Enter to exit...");
        input.nextLine();
    }

    public static void menu(Scanner input) {
        while (1 != 0) {
            clearScreen();
            System.out.println("======PROGRAM MENU======");
            System.out.println("1. Gas Station Program");
            System.out.println("2. Form a triangle");
            System.out.println("3. Salary increase in football");
            System.out.println("4. Grades higher than 9");
            System.out.println("5. Inverse sorter of numbers");
            System.out.println("6. Paying for parking");
            System.out.println("7. Wholesale discount");
            System.out.println("8. Cashier Program");
            System.out.println("9. Friendly Numbers");
            System.out.println("0. Exit");
            System.out.print("Select an option: ");
            int opt = input.nextInt();
            switch (opt) {
                case 0:
                    return;
                case 1:
                    gasStaion(input);
                    break;
                case 2:
                    triangle(input);
                    break;
                case 3:
                    salaryFoot(input);
                    break;
                case 4:
                    gradeHigh(input);
                    break;
                case 5:
                    inverNum(input);
                    break;
                case 6:
                    parking(input);
                    break;
                case 7:
                    descount(input);
                    break;
                case 8:
                    cashier(input);
                    break;
                case 9:
                    amiNum(input);
                    break;
                default:
                    break;
            }
        }

    }

    public static void gasStaion(Scanner input) {
        clearScreen();
        // Inputs block, saves in primitive variables with input object
        System.out.print("Enter the distance traveled in kilometers: ");
        double DKm = input.nextDouble();
        System.out.print("Enter the price per galon of gasoline: ");
        double PGas = input.nextDouble();
        System.out.print("Enter the total mount of money spent on gasoline: ");
        double MonGas = input.nextDouble();
        System.out.print("Enter the time in hours and minutes (use spaces to separate): ");
        double TimeH = input.nextDouble();
        double TimeM = input.nextDouble();
        // Calculation block
        double TimeT = TimeH + (TimeM / 60);
        double GalKm = (MonGas / PGas) / DKm;
        double PromKmH = DKm / TimeT;
        double PromMS = (DKm * 1000) / (TimeT * 3600);
        // Outputs Block
        System.out.printf("Spent %.2f galon per kilometer traveled%n", GalKm);
        System.out.printf("Its avarage speed was %.2fkm/h%n", PromKmH);
        System.out.printf("Its avarage speed was %.2fm/s%n", PromMS);
        pause(input);
        return;
    }

    public static void triangle(Scanner input) {
        clearScreen();
        double[] sides = new double[3];
        for (int i = 0; i < sides.length; i++) {
            System.out.printf("Enter the side %d of your triangle: ", i + 1);
            sides[i] = input.nextDouble();
        }
        Arrays.sort(sides);
        if ((sides[0] + sides[1]) > sides[2]) {
            System.out.println("The triangle can be formed");
        } else {
            System.out.println("The triangle can't be formed");
        }
        pause(input);
        return;
    }

    public static void salaryFoot(Scanner input) {
        clearScreen();
        double incre = 0;
        System.out.print("Enter the football player current salary: ");
        double salary = input.nextDouble();
        if (salary <= 6000) {
            incre = 20;
        } else if (salary <= 7900) {
            incre = 10;
        } else if (salary <= 12000) {
            incre = 5;
        } else {
            incre = 0;
        }
        double Fsalary = salary + (salary * (incre / 100));
        System.out.printf("Current Salary: %.2f %nIncrese: %f %nFinal Salary: %.2f", salary, incre, Fsalary);
        pause(input);
        return;
    }

    public static void gradeHigh(Scanner input) {
        clearScreen();
        int i = 0, j = 0;
        System.out.print("Enter the students they passed: ");
        int stds = input.nextInt();
        double[] Pstds = new double[stds];
        if (stds == 0) {
            System.out.println("Nobady passed");
            pause(input);
            return;
        }
        while (i != stds) {
            System.out.print("Enter the grades of the students who passed: ");
            double grade = input.nextDouble();
            if (grade > 10 || grade < 0) {
                System.out.println("Enter a valid grade");
            } else {
                Pstds[i] = grade;
                i++;
            }
        }
        for (i = 0; i < stds; i++) {
            if (Pstds[i] > 9) {
                j++;
            }
        }
        System.out.printf("The %.2f%% passed with more than 9", ((j * 100.0) / stds));
        pause(input);
        return;
    }

    public static void inverNum(Scanner input) { 
        String num;
        do {
            clearScreen(); // clearScreen() is a method created to clear the terminal before starting
            System.out.printf("Enter a 3-digit number: ");
            num = input.next(); // .next stores what was entered without line breaks and white spaces
            if (num.length() != 3) {
                System.out.println("Error: The number must have exactly 3 digits.");
                pause(input);
            }
        } while (num.length() != 3); // Validates that the number has 3 digits
        char[] numa = num.toCharArray(); // converts the previous string into a character array
        int j = num.length() - 1; // num.length() gives the length of the string, -1 is to match the index starting at 0
        int i = 0;
        while (i < j) {
            Character aux = numa[j];
            numa[j] = numa[i];
            numa[i] = aux; // The loop swaps the data according to i and j
            i++;
            j--;
        }
        num = new String(numa);
        System.out.printf("The inverted number is: %s", num);
        pause(input); // pause is a method created so the user can see the entire function before
                      // continuing
        return;
    }

    public static void parking(Scanner input) {
        boolean f = true;
        int time = 0;
        double pay = 0;
        while (f) { // The loop is made in case the user does not enter an established option using
                    // a boolean
            clearScreen();
            System.out.println(
                    "According to your vehicle, enter:\n'a' for a car\n'b' for a bus\n'c' for a truck");
            System.out.print("Enter your option: ");
            Character opt = input.next().charAt(0); // Stores the character that the user typed
            switch (opt) {
                case 'a':
                    pay = 1.20;
                    f = false; // Changes the boolean to false to be able to exit the loop
                    break;
                case 'b':
                    f = false;
                    pay = 2;
                    break;
                case 'c':
                    pay = 3;
                    f = false;
                    break;
                default:
                    System.out.println("Enter a correct option: ");
                    pause(input);
                    break;
            }
        }
        System.out.print("Enter the amount of time in hours that the vehicle was in the parking lot: ");
        time = input.nextInt();
        System.out.printf("You must pay %.2f$", (time * pay)); // %2.f shows two decimals of the result
        pause(input);
        return;
    }

    public static void descount(Scanner input) {
        int cant = 0;
        int desc = 0;
        double priceu = 0;
        while (cant == 0) { // verifies that something was purchased
            clearScreen();
            System.out.print("Enter the quantity of the product you are going to buy: ");
            cant = input.nextInt();
            if (cant == 0) {
                System.out.println("Add something to the cart: ");
                pause(input);
            }
        }
        while (priceu == 0) { // verifies that the price is not 0
            clearScreen();
            System.out.print("Enter the unit price of the product: ");
            priceu = input.nextDouble();
            if (priceu == 0) {
                System.out.print("Enter the correct price");
                pause(input);
            }
        }
        if (cant < 10) { // Will assign a discount depending on the quantity of items
            desc = 0;
        } else if (cant < 50) {
            desc = 3;
        } else if (cant < 100) {
            desc = 5;
        } else {
            desc = 10;
        }
        clearScreen();
        System.out.printf("Purchased items: %d%n", cant);
        System.out.printf("Unit price: %.2f$%n", priceu);
        System.out.printf("Discount based on the quantity of products: %d%%%n", (desc));
        System.out.printf("The total to pay is: %.2f$", ((cant * priceu) - ((cant * priceu) * (desc / 100.0))));
        pause(input);
        return;
    }

    public static void cashier(Scanner input) {
        clearScreen();
        int bills = 0;
        int money = 0;
        int[] dbills = { 100, 50, 20, 10, 5, 1 };
        while (money == 0) {
            System.out.print("Enter the amount of money you are going to withdraw (integer number): ");
            money = input.nextInt();
            if (money == 0) {
                System.out.println("Enter a valid integer number");
                pause(input);
            }
        }
        System.out.println("You will receive: ");
        for (int i = 0; i < 6; i++) {
            bills = money / dbills[i]; // stores the number of bills that can be withdrawn depending on the denomination
                                       // (dbills)
            money = money % dbills[i]; // stores the remainder of the division to be able to use a lower denomination
            System.out.print(bills > 0 ? String.format("%d bill(s) of %d%n", bills, dbills[i]) : "");
        }
        pause(input);
        return;
    }

    public static void amiNum(Scanner input) {
        clearScreen();
        int sum1 = 0;
        int sum2 = 0;
        ArrayList<Integer> div1 = new ArrayList<>();
        ArrayList<Integer> div2 = new ArrayList<>();
        // To make it simpler, I imported the ArrayList library
        System.out.print("Enter the first number: ");
        int num1 = input.nextInt();
        System.out.print("Enter the second number: ");
        int num2 = input.nextInt();
        for (int i = 1; i < num1; i++) {
            if (num1 % i == 0) {
                sum1 += i;
                div1.add(i); // Adds the divisors of num1 to the list
            }
        }
        for (int i = 1; i < num2; i++) {
            if (num2 % i == 0) {
                sum2 += i;
                div2.add(i);// Adds the divisors of num2 to the list
            }
        }
        System.out.printf("Divisors of %d: %s%n", num1, div1);
        System.out.printf("Sum of divisors of %d: %d%n", num1, sum1);
        System.out.printf("Divisors of %d: %s%n", num2, div2);
        System.out.printf("Sum of divisors of %d: %d%n", num2, sum2);
        if ((sum1 == num2) && (sum2 == num1)) {
            System.out.println("They are friendly numbers");
        } else {
            System.out.println("They are not friendly numbers");
        }
        pause(input);
        return;
    }
}